### DISCUSSION TITLE
Module v8 not found after moving api to custom path

### DISCUSSION THREAD

**Discussion by petute** [nextauthjs/next-auth#4332] [Help]
### Question 💬

Hello

I have encountered the error `Module not found: Can't resolve 'v8'` after moving my next-auth api from /api to /next-auth/api.

I have to move the api because /api is already taken by a rest-api in production.

I thought all I had to do was move the [...next-auth].ts into /pages/next-auth/api/auth and change the NEXTAUTH_URL to https://example.com/next-auth/api/auth. Did I forget something?

I have already seen #3469 but still can't figure out what I'm doing wrong.
Thank you for taking the time to answer my question.

### How to reproduce ☕️

This is my code.

`[...next-auth].ts`
```typescript
import NextAuth from "next-auth";
import CredentialsProvider from "next-auth/providers/credentials";

export default NextAuth({
  pages: {
    signIn: "/login",
    error: "/login", // This is somewhat of an unclean solution.
  },
  providers: [
    CredentialsProvider({
      id: "credentials",
      name: "credentials",
      credentials: {
        username: { label: "username", type: "text" },
        password: { label: "password", type: "password" },
      },
      async authorize(credentials) {
        const res = await fetch("https://localhost:8000/api/login", {
          method: "POST",
          body: JSON.stringify({
            username: credentials?.username,
            password: credentials?.password,
          }),
          headers: {
            "Content-Type": "application/json",
            Accept: "application/json",
          },
        });

        const user = await res.json();
        if (res.ok && user) {
          return user;
        } else {
          return Promise.reject(new Error("Invalid login credentials."));
        }
      },
    }),
  ],
  callbacks: {
    async session({ session, token }) {
      session.accessToken = token.accessToken;

      return session;
    },
    async jwt({ token, user }) {
      if (user) {
        token.accessToken = user.token;
      }

      return token;
    },
  },
});
```

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by ThangHuuVu** [comment]
I don't think you can move next-auth API like that. You can, however, do this instead: https://github.com/vercel/next.js/discussions/14634#discussioncomment-43744

**Reply by petute** [reply]
Sadly this didn't work for me. At this point we decided to not include next-auth in our project for the time being because it would require too many changes.

**Reply by petute** [reply]
But thank you for the fast reply :)

**Comment by NeutrinoXY** [comment]
Same error here. Use of rewrites instead of the use of NEXTAUTH_URL doesn't help: it also causes the page [...nextauth].js to be compiled after the compilation of any other page, which of course triggers the error earlier (i.e. before request time).

**Reply by ThangHuuVu** [reply]
Seems odd... AFAIK, next-auth should still live in the `/api` folder, and on `next.config.js` you'd write the rewrite rule. A minimal reproduction repo would be super helpful here, could you help spin up one? 🙏

**Reply by NeutrinoXY** [reply]
The error appears only if next-auth is not in the /api folder (I was using a /api/auth -> /auth rewrite).

With the other way (rewrite /auth -> /api/auth), there is no compilation error, but I'm having issues with the callbackUrl.
Using this rewrite rule, I need to set callbackUrl to /auth/callback/<provider> instead of default /api/auth/callback/<provider>, but the callbackUrl I specify as argument in signIn function is not used. However, the cookie "next-auth.callback-url" is correctly set to the custom value. 
Is the custom callbackUrl considered invalid since it doesn't live in /api/auth folder?

**Reply by NeutrinoXY** [reply]
After I discovered that callbackUrl is not used for the redirect_uri parameter in oauth2 authorization request, I overrode the provider authorization URL in order to set redirect_uri to /auth/callback/<provider>
Everything works fine then on provider side, but I receive a "redirect_uri_mismatch" error from openid-client library:

`[next-auth][error][OAUTH_CALLBACK_ERROR] `
`https://next-auth.js.org/errors#oauth_callback_error redirect_uri_mismatch (Bad Request) {`
`  error: {`
`    message: 'redirect_uri_mismatch (Bad Request)',`
`    stack: 'OPError: redirect_uri_mismatch (Bad Request)\n' +`
`      '    at processResponse (X/node_modules/openid-client/lib/helpers/process_response.js:38:13)\n' +`
`      '    at Client.grant (X/node_modules/openid-client/lib/client.js:1340:22)\n' +`
`      '    at runMicrotasks (<anonymous>)\n' +`
`      '    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n' +`
`      '    at async Client.callback (X/node_modules/openid-client/lib/client.js:474:24)\n' +`
`      '    at async oAuthCallback (X/node_modules/next-auth/core/lib/oauth/callback.js:112:16)\n' +`
`      '    at async Object.callback (X/node_modules/next-auth/core/routes/callback.js:50:11)\n' +`
`      '    at async NextAuthHandler (X/node_modules/next-auth/core/index.js:139:28)\n' +`
`      '    at async NextAuthNextHandler (X/node_modules/next-auth/next/index.js:21:19)\n' +`
`      '    at async X/node_modules/next-auth/next/index.js:57:32',`
`    name: 'OPError'`
`  },`
`  providerId: 'google',`
`  message: 'redirect_uri_mismatch (Bad Request)'`
`}`

It seems that the "rewrite fix" doesn't cover my (simple) use case.

**Comment by Webbrother** [comment]
The same problem apears when I do a very small refactoring. I move the function that uses `unstable_getServerSession` outside of `getServerSideProps`.


This is works fine: 

```typescript
export const getServerSideProps = createGetServerSidePropsWithParams<ChampPageProps, ChampPageQuery>(async context => {
  const { params, locale } = context;
  const { isAuthorized, config } = await getServerRequestConfig(context);

  try {
    return {
      props: {
        champDTO: await getChampDTOById(params.champId, isAuthorized, config),
        ...(await serverSideTranslations(locale, ['common', 'champ', 'messages'])),
      },
    };
  } catch (e) {
    return {
      notFound: true,
    };
  }
});
```

This DOESN'T work:

```typescript
export const getServerSideProps = createGetServerSidePropsWithParams<ChampPageProps, ChampPageQuery>(async context => {
  const { params, locale } = context;

  try {
    return {
      props: {
        champDTO: await getChampDTOById(params.champId, context), // <- I just put `getServerRequestConfig` inside of this function
        ...(await serverSideTranslations(locale, ['common', 'champ', 'messages'])),
      },
    };
  } catch (e) {
    return {
      notFound: true,
    };
  }
});
```

**Comment by socketopp** [comment]
@petute Did you resolve this issue eventually?

**Comment by petute** [comment]
@socketopp No. It was for a freelance gig and as stated I didn't deem it worth the time investment to fix their stuff. 😅

**Comment by gentlementlegen** [comment]
Had the same error while using `unstable_getServerSession` outside of a ServerSideProps (was using in Apollo to fetch my token). My first fix was to install [the v8 package directly](https://www.npmjs.com/package/v8). But very recently it disappeared for some reason.
My current solution is to call `getSession({ req })` instead, which seem to work. No sure if that solves your problem exactly, hopefully someone finds this information useful.
